gateway/src/apicast/policy/rate_limit/rate_limit.lua

```diff
@@ -0,0 +1,222 @@
+
+local resty_limit_conn = require('resty.limit.conn')
+local resty_limit_req = require('resty.limit.req')
+local resty_limit_count = require('resty.limit.count')
+
+local ngx_semaphore = require "ngx.semaphore"
+local limit_traffic = require "resty.limit.traffic"
+local ts = require ('apicast.threescale_utils')
+local tonumber = tonumber
+local next = next
+local shdict_key = 'limiter'
+
+local new = _M.new
+
+local traffic_limiters = {
+  connections = function(config)
+    return resty_limit_conn.new(shdict_key, config.conn, config.burst, config.delay)
+  end,
+  leaky_bucket = function(config)
+    return resty_limit_req.new(shdict_key, config.rate, config.burst)
+  end,
+  fixed_window = function(config)
+    return resty_limit_count.new(shdict_key, config.count, config.window)
+  end
+}
+
+local function try(f, catch_f)
+  local status, exception = pcall(f)
+  if not status then
+    catch_f(exception)
+  end
+end
+
+local function init_limiter(config)
+  local lim, limerr
+  try(
+    function()
+      lim, limerr = traffic_limiters[config.name](config)
+    end,
+    function(e)
+      return nil, e
+    end
+  )
+  if not lim then
+    return nil, limerr
+  end
+
+  return lim, nil
+end
+
+local function redis_shdict(url)
+  local options = { url = url }
+  local redis, err = ts.connect_redis(options)
+  if not redis then
+    return nil, err
+  end
+
+  return {
+    incr = function(_, key, value, init)
+      if not init then
+        return redis:incrby(key, value), nil
+      end
+      redis:setnx(key, init)
+      return redis:incrby(key, value), nil
+    end,
+    set = function(_, key, value)
+      return redis:set(key, value)
+    end,
+    expire = function(_, key, exptime)
+      local ret = redis:expire(key, exptime)
+      if ret == 0 then
+        return nil, "not found"
+      end
+      return true, nil
+    end,
+    get = function(_, key)
+      local val = redis:get(key)
+      if type(val) == "userdata" then
+        return nil
+      end
+      return val
+    end
+  }
+end
+
+local function error(logging_only, status_code)
+  if not logging_only then
+    return ngx.exit(status_code)
+  end
+end
+
+function _M.new(config)
+  local self = new()
+  self.config = config or {}
+  self.limiters = config.limiters
+  self.redis_url = config.redis_url
+  self.status_code_rejected = config.status_code_rejected or 429
+  self.logging_only = config.logging_only or false
+
+  return self
+end
+
+function _M:access()
+  local limiters = {}
+  local keys = {}
+
+  if not self.redis_url then
+    -- Only one (the first) limiter is enable.
+    -- Key will be shdict_key ('limiter').
+    local lim, initerr = init_limiter(self.limiters[1])
+    if not lim then
+      ngx.log(ngx.ERR, "unknown limiter: ", initerr)
+      error(self.logging_only, 500)
+      goto done
+    end
+
+    limiters[1] = lim
+    keys[1] = shdict_key
+
+  else
+    for _, limiter in ipairs(self.limiters) do
+      local lim, initerr = init_limiter(limiter)
```

Comment:
> I meant to move the limiter initialization to _M.new so they are created just once per policy.

@mikz If I move the limiter initialization to _M.new, "shared dict not found" error occurs. Maybe you know this behavior, so you put the limiter initialization to _M.access [here](https://github.com/3scale/conn-limiter/blob/master/src/conn-limiter.lua#L44), right?